src/clojure/nrepl/cmdline.clj

```diff
@@ -239,59 +244,193 @@
+            :middleware middleware
+            :handler (get-handler-option options middleware)
+            :greeting (get-greeting-option options transport)
+            :ack-port (get-ack-port options)})))
+
+(defn connect-command
+  "Connects to a running nREPL server and runs a REPL. Exits program when REPL
+  is closed.
+  Takes a map of nREPL CLI options."
+  [options]
+  (let [{:keys [host port transport]} (get-conn-options options)]
+    (run-repl host port {:transport transport})
+    (exit 0)))
+
+(defn ack-server
+  "Acknowledge the port of this server to another nREPL server running on
+  :ack port.
+  Takes nREPL server map and processed CLI options map.
+  Prints a message describing the acknowledgement between servers.
+  Returns nil."
+  [server options]
+  (when-let [ack-port (get-ack-port options)]
+    (let [port (:port server)
+          transport (:transport options)]
+      (binding [*out* *err*]
+        (println (format "ack'ing my port %d to other server running on port %d"
+                         port ack-port)
+                 (:status (send-ack port ack-port transport)))))))
+
+(defn print-connection-header
+  "Prints the nREPL server header that some tools rely on to parse the
+  connection details from.
+  Takes nREPL server map and processed CLI options map.
+  Returns nil."
+  [server options]
+  (let [transport (:transport options)
+        port (:port server)
+        ^java.net.ServerSocket ssocket (:server-socket server)
+        host (.getHostName (.getInetAddress ssocket))]
+    ;; The format here is important, as some tools (e.g. CIDER) parse the string
+    ;; to extract from it the host and the port to connect to
+    (println (format "nREPL server started on port %d on host %s - %s://%s:%d"
+                     port host (transport/uri-scheme transport) host port))))
+
+(defn save-port-file!
+  "Writes a file relative to project classpath with port number so other tools
+  can infer the nREPL server port.
+  Takes nREPL server map and processed CLI options map.
+  Returns nil."
+  [server options]
+  ;; Many clients look for this file to infer the port to connect to
+  (let [port (:port server)
+        port-file (io/file ".nrepl-port")]
+    (.deleteOnExit port-file)
+    (spit port-file port)))
+
+(defn interactive-repl-or-sleep
+  "Runs an interactive repl if :interactive CLI option is true otherwise
+  puts the current thread to sleep
+  Takes nREPL server map and processed CLI options map.
+  Returns nil."
+  [server options]
+  (if (:interactive options)
+    (let [transport (:transport options)
+          host (:host server)
+          port (:port server)]
+      (run-repl host port (merge (when (:color options) colored-output)
+                                 {:transport transport})))
+      ;; need to hold process open with a non-daemon thread
+      ;;   -- this should end up being super-temporary
+    (Thread/sleep Long/MAX_VALUE)))
+
+(defn server-command
```

Comment:
The name of this command seems misleading to me, given it actually starts the server and connects a REPL to it. I don't think it's a good idea to complect the both things.

Reply:
Plus splitting it up would be better for future custom CLI wrappers that want their own pipeline. I'll move the pipeline in that function into the main function along with the run logic as you mentioned above.